### PR TITLE
Default new program with project and Tasks

### LINKED ISSUE #49: Create Program API creates default projects and tasks
Related to #34

### PR DESCRIPTION
**Pull request by proserve**
Close #49 

Changes:
- Auto-populate created programs with default project and Tasks template
- as per discussion with @brettstack, Refactor partition key, remove dependency to user Id.
- Prepare postman API documentation to be used by FE developer, [find it here](https://www.getpostman.com/collections/8bb22c1d9e900982b71d)
- Code clean-up and API reorganization, Put project APIs under `/:programId`

# Checklist

- [ ] Tests have been added and are passing
- [x] Documentation has been updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/api/controllers/program.js [hunk 1]
```diff
@@ -42,21 +87,16 @@ export async function updateProgram({
   return programItem.Attributes
 }
 
-export function getPrograms({ userId }) {
-  if (!userId) throw new Error('userId is required')
-
-  return Program.query(userId)
+export async function getProgram({ programId, userId }) {
```
Comment (brettstack) [inline packages/api/controllers/program.js:90]: Please rename to `getProgramWithWorkspace`

File: packages/api/controllers/program.js [hunk 2]
```diff
@@ -42,21 +87,16 @@ export async function updateProgram({
...
-  if (!programId) throw new Error('programId is required')
-
-  const programItem = await Program.get({ userId, id: programId })
-
-  if (!programItem) {
-    return null
-  }
-
-  return programItem.Item
+export function scanPrograms(workspaceId) {
```
Comment (brettstack) [inline packages/api/controllers/program.js:100]: Please change this to `getPrograms({ workspaceId })`

File: packages/api/controllers/project.js [hunk 3]
```diff
@@ -12,10 +12,9 @@ export async function createProject({
   if (!project) throw new Error('project is required')
 
   const id = generateId()
-  const partitionKey = getPk({ userId, programId })
   const item = {
     ...project,
-    programKey: partitionKey,
+    programId,
     id,
```
Comment (brettstack) [inline packages/api/controllers/project.js:18]: So.... If it's not too much trouble.... Could we change all of these `id`s to (for example) `projectId`? I should have done this from the start.

File: packages/api/jest-dynalite-config.js [hunk 4]
```diff
@@ -33,31 +33,31 @@ module.exports = {
...
+      KeySchema: [{ AttributeName: 'programId', KeyType: 'HASH' }, { AttributeName: 'id', KeyType: 'RANGE' }],
+      AttributeDefinitions: [{ AttributeName: 'programId', AttributeType: 'S' }, { AttributeName: 'id', AttributeType: 'S' }],
       ProvisionedThroughput: {
         ReadCapacityUnits: 1,
         WriteCapacityUnits: 1,
       },
       data: [
         {
-          programKey: 'user1#program1',
+          programId: 'user1#program1',
```
Comment (brettstack) [inline packages/api/jest-dynalite-config.js:44]: Please change things like this which should no longer have the `user1#` prefix

File: packages/api/models/Task.js [hunk 5]
```diff
@@ -21,14 +20,14 @@ const Task = new Entity({
     name: 'string',
     description: 'string',
     status: 'string',
-    assignee: 'string',
-    budget: 'string',
+    assignee: 'list',
```
Comment (brettstack) [inline packages/api/models/Task.js:23]: Should this be `assignees`? @sanathkr

**Review by brettstack — CHANGES_REQUESTED** [review]
Brilliant work! I didn't see how we're limiting access to only those that have permission. Happy to do that as a follow-up PR. For that we should certainly have integ tests.

File: packages/api/controllers/program.js [hunk 6]
```diff
@@ -42,21 +87,16 @@ export async function updateProgram({
   return programItem.Attributes
 }
 
-export function getPrograms({ userId }) {
-  if (!userId) throw new Error('userId is required')
-
-  return Program.query(userId)
+export async function getProgram({ programId, userId }) {
+  const { Items } = await Program.scan({ filters: { attr: 'id', eq: programId } })
```
[the remark this answers is not in the document]
Comment (proserve) [reply packages/api/controllers/program.js:91]: @brettstack so all other API under `/program` will be moved to `/workspaces/:workspaceId`

File: packages/api/controllers/workspace.js [hunk 7]
```diff
@@ -3,7 +3,20 @@ import { generateId } from '../utils/id'
...
+    filters: [{
+      attr: 'isDefault',
+      eq: true,
+    }, {
+      attr: 'userId',
+      eq: userId,
+    }],
+  })
+  return defaultWorkspace.Items && defaultWorkspace.Items[0]
+}
```
[the remark this answers is not in the document]
Comment (proserve) [reply packages/api/controllers/workspace.js:19]: @brettstack if we move to create program API to `/workspaces/:workspaceId/program`, this query will not be used.

File: packages/api/controllers/project.js (hunk 3 above)
Comment (proserve) [reply packages/api/controllers/project.js:18]: I thought we made `programId` as a partition key for `Project` table just to make DynamoDb store the program's projects close to each other, which will make the fetch more efficient.

Comment (brettstack) [reply packages/api/controllers/project.js:18]: Yeah I'm just talking about the `id` field, rename it to the more explicit `projectId`

**Review by brettstack — APPROVED** [review]
This is great! We can do the `id` => `*Id` renaming in a separate PR.

**Comment by sanathkr** [comment]
It always comes down to IDs ;)

**Pull request merged**